### ISSUE TITLE
Problem with SignalReceived activity

### ISSUE THREAD

**Issue by jkixj** [elsa-workflows/elsa-core#3274]
I posted this in Discord and was asked to post it here.

I am using Elsa 2.8.2 and I am embedding it in an ASP.Net application running under .Net 6. The application runs locally under IIS Express.

I created a test workflow that exhibits the problem:
```csharp
public class SignalTestCase : IWorkflow
{
    public void Build(IWorkflowBuilder builder)
    {
        builder.HttpEndpoint("/signal-test")
            .SetVariable("ApproveSignal", c => c.GenerateSignalUrl("ApproveSignal"))
            .SetVariable("RejectSignal", c => c.GenerateSignalUrl("RejectSignal"))
            .WriteHttpResponse((activity) => activity
                .WithStatusCode(HttpStatusCode.OK)
                .WithContentType("application/json")
                .WithContent((context) =>
                {
                    var approveSignalUrl = context.GetVariable<string>("ApproveSignal");
                    var rejectSignalUrl = context.GetVariable<string>("RejectSignal");
                    return new { approveSignalUrl, rejectSignalUrl };
                }))
            .Then<Fork>(setup: activity => activity.WithBranches("Approve", "Reject")
                , fork =>
                {
                    fork.When("Approve")
                        .SignalReceived("ApproveSignal").WithId("ApproveSignalReceived")
                        .SetVariable("Status", "Approved")
                        .ThenNamed("Join");
                    fork.When("Reject")
                        .SignalReceived("RejectSignal").WithId("RejectSignalReceived")
                        .SetVariable("Status", "Rejected")
                        .ThenNamed("Join");
                })
            .Add<Join>(activity => activity.WithMode(Join.JoinMode.WaitAny)).WithName("Join")
            .WriteHttpResponse(activity => activity.WithStatusCode(HttpStatusCode.OK)
                .WithContentType("text/plain")
                .WithContent(x => x.GetVariable("Status"))
            )
            .Finish("Complete");
    }
}
```

My Elsa Server has the following settings:
```json
"Server": {
      "BaseUrl": "http://localhost:38888",
      "BasePath": "/http"
}
```

When I hit the url `http://localhost:38888/http/signal-test`, I get the expected response:
```json
{
    "approveSignalUrl": "http://localhost:38888/signals/trigger/CfDJ8NoaAM8dbixKpFhnzYe5HUUa5QGL8p4fqqYdGatlVyeU7QHrfXaMTbYiEtR9E1H4Dot9AuAkVAW7aAqirBxT3QYpiO0JnPydOFHq6_ad80l90gIEiXbckllxOZ9ZlPpfBucQe7FR6tA0RtBHyRTNrAVGlE1o54GfiYBE1K5PSCTmqion_yov6nylFgJpMlrQ8-C02lZin04Ri63R8d7aM1p_6keyrtevIYgN_q9lo5ta",
    "rejectSignalUrl": "http://localhost:38888/signals/trigger/CfDJ8NoaAM8dbixKpFhnzYe5HUVC8Z-fi6dUcoDeGO8WycX8ux7GVJEvbx_JmK7ykuLWqJ5X5iWl4yDlaH_Hj8tQcrKh5q8ii6K_Y99JA0Kk3uDDxsZkhF8umcjL7hs5-qUmgCDJDTeOmd8UhvPt2wjumtf0yxzwwr9FG_M4G_JkFytHQ-9m1LoVhfA3ITnmrxrxuyklVJHSfgVpInDDmuLHhow"
}
```

Then, if I access the approveSignalUrl, I the response:
```json
[]
```
I think that indicates that no workflows were affected?

I can then use the Swagger UI to interface with the Elsa server (I can't get the Dashboard to work, but that's a separate issue).

If I request a list of the instances I get:
```json
{
  "$id": "1",
  "items": [
    {
      "$id": "2",
      "id": "379327161fed4b70b4561c112d1204f9",
      "definitionId": "SignalTestCase",
      "definitionVersionId": "SignalTestCasev1",
      "version": 1,
      "workflowStatus": "Suspended",
      "correlationId": "12f8292c84ef4e0c9628cce5d669248f",
      "createdAt": "2022-08-25T16:57:26.5281228Z",
      "lastExecutedAt": "2022-08-25T16:57:26.7925397Z",
      "metadata": {
        "$id": "3"
      }
    }
  ],
  "page": 0,
  "pageSize": 25,
  "totalCount": 1
}
```

And if I ask for the details of that instance, I get:
```json
{
  "$id": "1",
  "definitionId": "SignalTestCase",
  "definitionVersionId": "SignalTestCasev1",
  "version": 1,
  "workflowStatus": "Suspended",
  "correlationId": "12f8292c84ef4e0c9628cce5d669248f",
  "createdAt": "2022-08-25T16:57:26.5281228Z",
  "lastExecutedAt": "2022-08-25T16:57:26.7925397Z",
  "variables": {
    "$id": "2",
    "data": {
      "ApproveSignal": "http://localhost:38888/signals/trigger/CfDJ8NoaAM8dbixKpFhnzYe5HUUa5QGL8p4fqqYdGatlVyeU7QHrfXaMTbYiEtR9E1H4Dot9AuAkVAW7aAqirBxT3QYpiO0JnPydOFHq6_ad80l90gIEiXbckllxOZ9ZlPpfBucQe7FR6tA0RtBHyRTNrAVGlE1o54GfiYBE1K5PSCTmqion_yov6nylFgJpMlrQ8-C02lZin04Ri63R8d7aM1p_6keyrtevIYgN_q9lo5ta",
      "RejectSignal": "http://localhost:38888/signals/trigger/CfDJ8NoaAM8dbixKpFhnzYe5HUVC8Z-fi6dUcoDeGO8WycX8ux7GVJEvbx_JmK7ykuLWqJ5X5iWl4yDlaH_Hj8tQcrKh5q8ii6K_Y99JA0Kk3uDDxsZkhF8umcjL7hs5-qUmgCDJDTeOmd8UhvPt2wjumtf0yxzwwr9FG_M4G_JkFytHQ-9m1LoVhfA3ITnmrxrxuyklVJHSfgVpInDDmuLHhow"
    }
  },
  "input": {
    "$id": "3",
    "providerName": "BlobStorage"
  },
  "output": {
    "$id": "4",
    "activityId": "activity-RejectSignalReceived"
  },
  "activityData": {
    "$id": "5",
    "activity-1": {
      "$id": "6",
      "_Lifecycle": {
        "$id": "7",
        "$type": "Elsa.Models.ActivityLifecycle, Elsa.Core",
        "executing": true,
        "executed": true
      }
    },
    "activity-2": {
      "$id": "8",
      "_Lifecycle": {
        "$id": "9",
        "$type": "Elsa.Models.ActivityLifecycle, Elsa.Core",
        "executing": true,
        "executed": true
      }
    },
    "activity-3": {
      "$id": "10",
      "_Lifecycle": {
        "$id": "11",
        "$type": "Elsa.Models.ActivityLifecycle, Elsa.Core",
        "executing": true,
        "executed": true
      }
    },
    "activity-4": {
      "$id": "12",
      "_Lifecycle": {
        "$id": "13",
        "$type": "Elsa.Models.ActivityLifecycle, Elsa.Core",
        "executing": true,
        "executed": true
      }
    },
    "activity-5": {
      "$id": "14",
      "_Lifecycle": {
        "$id": "15",
        "$type": "Elsa.Models.ActivityLifecycle, Elsa.Core",
        "executing": true,
        "executed": true
      }
    },
    "activity-ApproveSignalReceived": {
      "$id": "16",
      "_Lifecycle": {
        "$id": "17",
        "$type": "Elsa.Models.ActivityLifecycle, Elsa.Core",
        "executing": true,
        "executed": true
      }
    },
    "activity-RejectSignalReceived": {
      "$id": "18",
      "_Lifecycle": {
        "$id": "19",
        "$type": "Elsa.Models.ActivityLifecycle, Elsa.Core",
        "executing": true,
        "executed": true
      }
    }
  },
  "metadata": {
    "$id": "20"
  },
  "blockingActivities": [
    {
      "$id": "21",
      "activityId": "activity-ApproveSignalReceived",
      "activityType": "SignalReceived"
    },
    {
      "$id": "22",
      "activityId": "activity-RejectSignalReceived",
      "activityType": "SignalReceived"
    }
  ],
  "scheduledActivities": [],
  "scopes": [],
  "lastExecutedActivityId": "activity-RejectSignalReceived",
  "id": "379327161fed4b70b4561c112d1204f9"
}
```

As you can see, even after I tried to signal, the instance is blocked on the approve/reject fork.

Back in my server logs, I see this from when I triggered the signal:
```
2022-08-25 12:58:22.489 -04:00 [INF] Request starting HTTP/1.1 GET http://localhost:38888/signals/trigger/CfDJ8NoaAM8dbixKpFhnzYe5HUUa5QGL8p4fqqYdGatlVyeU7QHrfXaMTbYiEtR9E1H4Dot9AuAkVAW7aAqirBxT3QYpiO0JnPydOFHq6_ad80l90gIEiXbckllxOZ9ZlPpfBucQe7FR6tA0RtBHyRTNrAVGlE1o54GfiYBE1K5PSCTmqion_yov6nylFgJpMlrQ8-C02lZin04Ri63R8d7aM1p_6keyrtevIYgN_q9lo5ta - -
2022-08-25 12:58:22.491 -04:00 [INF] Executing endpoint 'Elsa.Activities.Http.Endpoints.Signals.TriggerEndpoint.Handle (Elsa.Activities.Http)'
2022-08-25 12:58:22.493 -04:00 [INF] Route matched with {action = "Handle", controller = "TriggerEndpoint"}. Executing controller action with signature System.Threading.Tasks.Task`1[Microsoft.AspNetCore.Mvc.IActionResult] Handle(System.String, System.Threading.CancellationToken) on controller Elsa.Activities.Http.Endpoints.Signals.TriggerEndpoint (Elsa.Activities.Http).
2022-08-25 12:58:22.507 -04:00 [INF] Executed DbCommand (2ms) [Parameters=[@__ActivityType_1='?' (Size = 450), @__Hash_2='?' (Size = 450), @__p_3='?' (DbType = Int32), @__p_4='?' (DbType = Int32)], CommandType='"Text"', CommandTimeout='30']
SELECT [b].[Id], [b].[ActivityId], [b].[ActivityType], [b].[CorrelationId], [b].[Hash], [b].[Model], [b].[ModelType], [b].[TenantId], [b].[WorkflowInstanceId]
FROM [Elsa].[Bookmarks] AS [b]
WHERE (([b].[TenantId] IS NULL) AND ([b].[ActivityType] = @__ActivityType_1)) AND ([b].[Hash] = @__Hash_2)
ORDER BY [b].[Id]
OFFSET @__p_3 ROWS FETCH NEXT @__p_4 ROWS ONLY
2022-08-25 12:58:22.514 -04:00 [INF] Executing OkObjectResult, writing value of type 'System.Collections.Generic.List`1[[Elsa.Services.Models.CollectedWorkflow, Elsa.Abstractions, Version=2.8.2.0, Culture=neutral, PublicKeyToken=null]]'.
2022-08-25 12:58:22.515 -04:00 [INF] Executed action Elsa.Activities.Http.Endpoints.Signals.TriggerEndpoint.Handle (Elsa.Activities.Http) in 21.7669ms
2022-08-25 12:58:22.515 -04:00 [INF] Executed endpoint 'Elsa.Activities.Http.Endpoints.Signals.TriggerEndpoint.Handle (Elsa.Activities.Http)'
2022-08-25 12:58:22.516 -04:00 [INF] Request finished HTTP/1.1 GET http://localhost:38888/signals/trigger/CfDJ8NoaAM8dbixKpFhnzYe5HUUa5QGL8p4fqqYdGatlVyeU7QHrfXaMTbYiEtR9E1H4Dot9AuAkVAW7aAqirBxT3QYpiO0JnPydOFHq6_ad80l90gIEiXbckllxOZ9ZlPpfBucQe7FR6tA0RtBHyRTNrAVGlE1o54GfiYBE1K5PSCTmqion_yov6nylFgJpMlrQ8-C02lZin04Ri63R8d7aM1p_6keyrtevIYgN_q9lo5ta - - - 200 2 application/json;+charset=utf-8 26.6165ms
```

Curious about the SQL query in the log, I check and find zero rows in the `Bookmarks` table.

In the triggers table I do find a relevant record:

|Column|Value
|------|------
|Id|eebc1b55d0994f8593c5be802bdddeb7
|TenantId|NULL
|Hash|f77b79f2512877f929d237bf90d8adc002bce4ec7545d1fb84bc5d5cecfaa715
|Model|{"Path":"/signal-test","Method":"get"}
|ModelType|Elsa.Activities.Http.Bookmarks.HttpEndpointBookmark, Elsa.Activities.Http
|ActivityType|HttpEndpoint
|ActivityId|activity-1
|WorkflowDefinitionId|SignalTestCase

Here are the Elsa pieces of my Startup.cs:
```csharp
string connectionString = _configuration.GetConnectionString(DefaultConnectionStringName);
var blobStoreConnectionString = _configuration.GetConnectionString(BlobStorageConnectionStringName);
var workflowAssemblyPath = Path.GetDirectoryName(typeof(HelloWorkflow).Assembly.Location);
services.Configure<BlobStorageWorkflowProviderOptions>(options =>
    options.BlobStorageFactory = () =>
        StorageFactory.Blobs.DirectoryFiles(Path.Combine(workflowAssemblyPath!, "Workflows")));
services.Configure<BlobStorageWorkflowStorageProviderOptions>(options =>
{
    options.BlobStorageFactory = () => StorageFactory.Blobs.FromConnectionString(blobStoreConnectionString);
});
services.AddSingleton<AuthenticationBasedHttpEndpointAuthorizationHandler>();
services.AddSingleton<EntityFrameworkBookmarkStore>();
services.AddElsa(elsa =>
    {
        elsa.WithContainerName("MyApp")
            .UseEntityFrameworkPersistence(ef =>
            {
                var migrationsAssembly =
                    typeof(global::Elsa.Persistence.EntityFramework.SqlServer.SqlServerElsaContextFactory)
                        .Assembly;
                ef.UseSqlServer(connectionString,
                    opts => opts.MigrationsAssembly(migrationsAssembly.GetName().Name)
                );
            })
            .UseDefaultWorkflowStorageProvider<BlobStorageWorkflowStorageProvider>()
            .UseBookmarkStore(sp => sp.GetRequiredService<EntityFrameworkBookmarkStore>())
            .ConfigureDistributedLockProvider(builder =>
            {
                // Make Elsa keep locks in the database (not memory)
                builder.UseSqlServerLockProvider(connectionString);
            })
            .AddConsoleActivities()
            .AddHttpActivities(options =>
            {
                _configuration.GetSection("Elsa:Server").Bind(options);
                options.HttpEndpointAuthorizationHandlerFactory = sp =>
                    sp.GetRequiredService<AuthenticationBasedHttpEndpointAuthorizationHandler>();
                options.HttpEndpointWorkflowFaultHandlerFactory = sp =>
                    sp.GetRequiredService<DefaultHttpEndpointWorkflowFaultHandler>();
            })
            .AddCommonTemporalActivities()
            .AddQuartzTemporalActivities(configureQuartz: configurator =>
            {
                // Make Quartz persist its state in the database (not memory)
                configurator.UsePersistentStore(store =>
                {
                    store.UseJsonSerializer();
                    store.UseSqlServer(options =>
                    {
                        options.ConnectionString = connectionString;
                        options.TablePrefix = "quartz.QRTZ_";
                    });
                    store.UseClustering();
                });
            })
            .AddJavaScriptActivities()
            .AddEmailActivities(options =>
            {
                _configuration.GetSection("Elsa:Smtp").Bind(options);
                if (!string.IsNullOrEmpty(options.PickupDirectoryLocation)
                    && !Directory.Exists(options.PickupDirectoryLocation))
                {
                    Directory.CreateDirectory(options.PickupDirectoryLocation);
                }
            })
            .AddFileActivities()
            .AddSqlServerActivities()
            .AddUserTaskActivities()
            .AddBlobStorageActivities()
            // ...other code adding custom workflow context providers, activities, and workflows
            ;
    })
    .AddLiquidExpressionEvaluator()
    .AddJavaScriptExpressionEvaluator()
    .WithJavaScriptOptions(options =>
    {
        // See https://elsa-workflows.github.io/elsa-core/docs/next/expressions/expressions-javascript#getconfig
        options.EnableConfigurationAccess = true;
    })
    .AddNotificationHandlersFrom<MyApp.Workflows.Scripting.ConfigureLiquidEngine>()
    .AddStartupRunner()
    .AddElsaSwagger()
    .AddElsaApiEndpoints();
```

I am not sure if this is a bug or me doing it incorrectly.

Thank you for any help in advance.

**Comment by jkixj** [comment]
Okay I think  got it working! I had this in my Startup.cs:
```csharp
.UseDefaultWorkflowStorageProvider<BlobStorageWorkflowStorageProvider>()
```
I used the Elsa.Samples.SignalingConsole project as a starting point and kept making it more similar to mine by adding a piece at a time into it. It stopped working when I added in that line. Now if I replace that line with:
```csharp
.UseDefaultWorkflowStorageProvider<WorkflowInstanceWorkflowStorageProvider>()
```
then it appears to work!

**Comment by jkixj** [comment]
I am closing this issue since it turned out to be an issue with my configuration.
